content/blog/2023-12-11-hlc/index.md

```diff
@@ -0,0 +1,538 @@
+Unlike the AST, it's flattened (no nested expressions or statements)
+and divided up into basic blocks, but unlike Caiman Assembly 
+it has traditional (non-CPS) control flow, and the inputs and outputs of
+funclets aren't explicit. One thing to note is that HIR function calls are considered
+basic block terminators, since in CPS a function call must be the last
+statement in a function. Furthermore, CFGs are canonicalized so that they have 
+a single entry and a single exit block.
+
+Once the AST is in CFG form, we perform a few local transformations as "preprocessing".
+Namely, we convert all operators to their typed forms 
+(so `add` becomes `_add_i64_i64` for adding two `i64`s) and convert all
+mutations of variables to loads and stores of a reference.
+
+One could imagine employing the progressive lowering ideas of CompCert and MLIR
+to separate the HIR into two HIRs, one before these transformations and
+a second representation afterward. I didn't go this route because it would
+involve decent code duplication for the definitions of the two IRs.
+
+# Implementation
+
+As an example, consider the following (toy) example in the high-level front-end:
+
+```
+// value specification
+val main() -> i64 {
+    b :- false
+    c :- true
+    d :- false
+    one :- 1
+    two :- 2
+    r :- c if b else d
+    z :- two if r else one
+    returns z
+}
+
+// implementation
+fn foo() -> i64 @ node(main.z)-usable 
+    impls main, time, space
+    // adheres to three specifications:
+    // * main  (value)
+    // * time  (timeline) (not shown)
+    // * space (spatial) (not shown)
+{
+    let b: bool @ node(main.b) = false;
+    var r: bool @ none(main);
+    if @ node(main.r) b {
+        let c: bool @ node(main.c) = true;
+        r = c;
+    } else {
+        let d: bool @ node(main.d) = false;
+        r = d;
+    }
+    @in { r: [node(main.r)-usable, none(space)-save] };
+    
+    var v: i64 @ none(main);
+    if @ node(main.z) r {
+        let two: i64 @ node(main.two) = 2;
+        v = two;
+    } else {
+        let one: i64 @ node(main.one) = 1;
+        v = one;
+    }
+    @in { v: [node(main.z)-usable, none(space)-save] };
+    return v;
+}
+```
+
+It's still not pretty on account of the current need to match pieces of 
+the implementation back to the spec almost everywhere (denoted with the `@`) 
+and keep expressions unnested. However, the schedule in this example has the following
+structure:
+
+<img src="hlc_1.svg" 
+    style="max-height: 50vh; display: block; margin-left: auto; margin-right: auto" />
+
+
+### CFG Generation
+
+During CFG generation, one piece of information we need to keep around is what
+the continuation of each basic block is. In the above example, the
+continuation of `foo` is `foo2` and the continuation of `foo2` is `foo3`.
+`foo4`, `foo5`, `foo6`, and `foo7` are at the maximum depth in this example, so
+when they're done, control is going to pass back to their parent's continuation.
```

Comment:
Based on the examples, am I correct in concluding that the continuation for a given block $b$ is the "join point" following $b$? Or, more formally, that it's $b$'s immediate post-dominator?